### ISSUE TITLE
Support for conditional default values

### ISSUE THREAD

**Issue by Reshad-Rahman** [netmod-wg/yang-next#112]
https://mailarchive.ietf.org/arch/msg/netmod/H1n_eARPuYz2D-Gu9IgzV86rQIg/

**Comment by abierman** [comment]
My suggestion was for the use-case where the default values are static.

    typedef def100 {
        type int32;
        default 100;
    }


    typedef def1000 {
        type int32;
        default 1000;
    }

    typedef defleaf {
        type union {
            type def100;
            type def1000;
        }
     }

Then a server can set the YANG default to 100 or 1000.
The client needs to use with-defaults=report-all to see what the server picked

**Comment by lllyfeng** [comment]
I think a optional default value or dynamic default value, and even fixed default value are all data generated by system. Maybe we should provide a rule to describe how system generated data appear or disappear, it can be yang extension written in yang files or  be a yang data file.

**Comment by rgwilton** [comment]
I don't think that we should have default values where the server can choose which value to use because this could mean that if a client also compiled the schema then it could end up with different default behaviour, and hence different semantics.

**Comment by abierman** [comment]
This seems to be a duplicate of #93 

This is too complicated to specify and to implement.
Interactions between when-stmt and dynamic defaults make it too difficult to implement and also
have all servers be consistent.

Real servers do add data to transactions (not just templates).
Sometimes dynamic defaults are needed based on conditions that are hard or impossible
to express in XPath.  Sometimes the server adds data in a completely different data structure.
This is a feature to make configuration simpler for the client.

Close

**Comment by janlindblad** [comment]
I think the basic principle is that if a client wishes to let a server decide about something, it should leave it unset. The description could specify the exact details of the server behavior for that case. Then if the client wishes to intervene, it can simply set/override the configs it needs to. Dynamic defaults open up many possibilities for interop issues.

**Comment by reshad-equinix** [comment]
> I don't think that we should have default values where the server can choose which value to use because this could mean that if a client also compiled the schema then it could end up with different default behaviour, and hence different semantics.

The case I had in mind (default MTU based on if-type) can not lead to different default behaviour. But I agree other scenarios may. I also agree with comments above wrt complexity. I'm ok with closing this.

**Comment by kwatsen** [comment]
Importance low (same as for #93)

For Andy's `union` example, does #72 resolve it?
  ^-- wait, this is a catch-22, since the data node wouldn't appear in the encoding

**Comment by BalazsLengyel** [comment]
There can be more default values or even e.g. temperature dependent defaults. IMHO the system should just set the value, that is allowed.
Such corner cases should be described in the description. We never wanted to do formal modelling of 100% of all constructs.

**Comment by reshad-equinix** [comment]
Yes this is usually tackled in the description. One annoying thing though is that if the node isn't mandatory and doesn't have a default, it is sometimes flagged as "unknown value". A new tag default-system would help but low-priority and as mentioned above I'm ok with closing this.
